okhttp/src/main/kotlin/okhttp3/MultipartReader.kt

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import java.io.Closeable
+import java.io.IOException
+import java.net.ProtocolException
+import okhttp3.internal.http1.HeadersReader
+import okio.Buffer
+import okio.BufferedSource
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import okio.Options
+import okio.Source
+import okio.Timeout
+import okio.buffer
+
+/**
+ * Reads a stream of [RFC 2046][rfc_2046] multipart body parts. Callers read parts one-at-a-time
+ * until [nextPart] returns null. After calling [nextPart] any preceding parts should not be read.
+ *
+ * Typical use loops over the parts in sequence:
+ *
+ * ```
+ * val multipartReader = MultipartReader(
+ *     boundary = "...",
+ *     source = source
+ * )
+ *
+ * multipartReader.use {
+ *   while (true) {
+ *     val part = multipartReader.nextPart() ?: break
+ *     process(part.headers, part.body)
+ *   }
+ * }
+ * ```
+ *
+ * Note that [nextPart] will skip any unprocessed data from the preceding part. If the preceding
+ * part is particularly large or if the underlying source is particularly slow, the [nextPart] call
+ * may be slow!
+ *
+ * Closing a part **does not** close this multipart reader; callers must explicitly close this with
+ * [close].
+ *
+ * [rfc_2046]: http://www.ietf.org/rfc/rfc2046.txt
+ */
+class MultipartReader(
+  private val source: BufferedSource,
+  @get:JvmName("boundary") val boundary: String
+) : Closeable {
+  /** This delimiter typically precedes the first part. */
+  private val dashDashBoundary = Buffer()
```

Comment:
Shouldn't these be defined on a companion object to avoid allocating and executing on each instance?